File: April_2022/08_04_2022/Color_of_building_blocks.cpp
#include <iostream>
#include <vector>

using namespace std;

int main(){

    int n,m;
    string s;

    vector<char> c;

    cin>>n;
    for(int i=0;i<n;i++){
        char temp;
        cin>>temp;
        c.push_back(temp);
    }

    cin>>m;
    cin>>s;

    vector<vector<int>> last_seen;
    
    for(int i=0;i<n;i++){
        vector<int> temp = {(int)c[i]};
        last_seen.push_back(temp);
    }

    for(int i=0;i<s.length();i++){
        for(int j=0;j<n;j++){
            if(last_seen[j][0] == (int)s[i]){
                last_seen[j].push_back(i);
                break;
            }
        }
    }

    for(int i=0;i<n;i++){
        for(int j=1;j<last_seen[i].size()-1;j++){
            if(abs(last_seen[i][j]-last_seen[i][j+1]) <= m){
                cout<<"Invalid";
                exit(0);
            }
        }
    }
    cout<<"Valid"<<endl;

    return 0;
}

File: April_2022/08_04_2022/Color_of_building_blocks.py
n = int(input())
c = input().split()
m = int(input())
s = input()
flag = False
last_seen = {}

for i in c:
    last_seen[i] = []

for i in range(len(s)):
    last_seen[s[i]].append(i)

for j in last_seen:
    for k in range(len(last_seen[j])-1):
        if(abs(last_seen[j][k]-last_seen[j][k+1])<=m):
            flag = True

if(flag):
    print("Invalid")
else:
    print("Valid")